Sources/SwiftParser/Attributes.swift

```diff
@@ -869,6 +876,152 @@ extension Parser {
   }
 }
 
+extension Parser {
+  mutating func parsePackageAttribute() -> RawAttributeSyntax {
+    let (unexpectedBeforeAtSign, atSign) = self.expect(.atSign)
+    let (unexpectedBeforePackageToken, packageToken) = self.expect(.keyword(._package))
+    let (unexpectedBeforeLeftParen, leftParen) = self.expect(.leftParen)
+    let arguments = self.parsePackageAttributeArguments()
+    let (unexpectedBeforeRightParen, rightParen) = self.expect(.rightParen)
+    return RawAttributeSyntax(
+      unexpectedBeforeAtSign,
+      atSignToken: atSign,
+      unexpectedBeforePackageToken,
+      attributeName: RawTypeSyntax(RawSimpleTypeIdentifierSyntax(name: packageToken, genericArgumentClause: nil, arena: self.arena)),
+      unexpectedBeforeLeftParen,
+      leftParen: leftParen,
+      argument: .packageAttributeArguments(arguments),
+      unexpectedBeforeRightParen,
+      rightParen: rightParen,
+      arena: self.arena
+    )
+  }
+
+  enum PackageLocationLabel: RawTokenKindSubset {
+    case id
+    case path
+    case url
+
+    init?(lexeme: Lexer.Lexeme) {
+      switch lexeme {
+      case RawTokenKindMatch(.id): self = .id
+      case RawTokenKindMatch(.path): self = .path
+      case RawTokenKindMatch(.url): self = .url
+      default: return nil
+      }
+    }
+
+    var rawTokenKind: RawTokenKind {
+      switch self {
+      case .id: return .keyword(.id)
+      case .path: return .keyword(.path)
+      case .url: return .keyword(.url)
+      }
+    }
+  }
+
+  enum PackageRequirementLabel: RawTokenKindSubset {
+    case branch
+    case from
+    case revision
+
+    init?(lexeme: Lexer.Lexeme) {
+      switch lexeme {
+      case RawTokenKindMatch(.branch): self = .branch
+      case RawTokenKindMatch(.from): self = .from
+      case RawTokenKindMatch(.revision): self = .revision
+      default: return nil
+      }
+    }
+
+    var rawTokenKind: RawTokenKind {
+      switch self {
+      case .branch: return .keyword(.branch)
+      case .from: return .keyword(.from)
+      case .revision: return .keyword(.revision)
+      }
+    }
+  }
+
+  mutating func parsePackageAttributeArguments() -> RawPackageAttributeArgumentsSyntax {
+    // Parsing package location.
+    let locationLabel = self.consume(ifAnyIn: PackageLocationLabel.self) ?? missingToken(.identifier)
+    let (unexpectedBeforeLocationColon, locationColon) = self.expect(.colon)
+    let (unexpectedBeforeLocation, location) = self.expect(.stringLiteral)
+    // Parsing package requirement.
+    let (unexpectedBeforeLocReqComma, locReqComma): (RawUnexpectedNodesSyntax?, RawTokenSyntax?)
```

Comment:
I don’t think you need these type annotations.

(same below)

Reply:
It seems they're still required. I'll extract the parsing logic to independent methods to keep this one clean. WIP